java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java

```diff
@@ -0,0 +1,243 @@
+      case HIVE:
+        return DateTimeFormat.forPattern(granularityType.getHiveFormat());
+      case LOWER_DEFAULT:
+        return DateTimeFormat.forPattern(granularityType.getLowerDefaultFormat());
+      default:
+        throw new IAE("There is no format for type %s", type);
+    }
+  }
+
+  @Override
+  public DateTime increment(DateTime time)
+  {
+    return new DateTime(periodGranularity.increment(time.getMillis()));
+  }
+
+  @Override
+  public DateTime decrement(DateTime time)
+  {
+    return new DateTime(periodGranularity.decrement(time.getMillis()));
+  }
+
+  @Override
+  public DateTime truncate(DateTime time)
+  {
+    return new DateTime(periodGranularity.truncate(time.getMillis()));
+  }
+
+  @Override
+  public DateTime toDate(String filePath, Formatter formatter)
+  {
+    Integer[] vals = getDateValues(filePath, formatter);
+
+    DateTime date = GranularityType.getDateTime(granularityType, vals);
+
+    if (date != null) {
+      return truncate(date);
+    }
+
+    return null;
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    PeriodSegmentGranularity that = (PeriodSegmentGranularity) o;
+
+    if (granularityType != that.granularityType) {
+      return false;
+    }
+    return periodGranularity.equals(that.periodGranularity);
+
+  }
+
+  @Override
+  public int hashCode()
+  {
+    int result = granularityType.hashCode();
+    result = 31 * result + periodGranularity.hashCode();
+    return result;
+  }
+
+  @Override
+  public String toString()
+  {
+    return "{type=period, " +
+           "period=" + getPeriod() +
+           ", timeZone=" + getTimeZone() +
+           ", origin=" + getOrigin() +
+           '}';
+  }
+
+  @Override
+  public void serialize(JsonGenerator jsonGenerator, SerializerProvider serializerProvider)
+      throws IOException, JsonProcessingException
+  {
+    // Retain the same behavior as pre-refactor granularity code.
+    // i.e. when Granularity class was an enum.
+    if (equalsPredefinedGranularities()) {
+      jsonGenerator.writeString(granularityType.toString());
+    } else {
+      jsonGenerator.writeStartObject();
+      jsonGenerator.writeStringField("type", "period");
+      jsonGenerator.writeObjectField("period", getPeriod());
+      jsonGenerator.writeObjectField("timeZone", getTimeZone());
+      jsonGenerator.writeObjectField("origin", getOrigin());
+      jsonGenerator.writeEndObject();
+    }
+  }
+
+  @Override
+  public void serializeWithType(
+      JsonGenerator jsonGenerator,
+      SerializerProvider serializerProvider,
+      TypeSerializer typeSerializer
+  ) throws IOException, JsonProcessingException
+  {
+    serialize(jsonGenerator, serializerProvider);
+  }
+
+  private boolean equalsPredefinedGranularities() {
+    final GranularityType[] values = GranularityType.values();
```

Comment:
With the currently suggested code, if a new predefined granularity added, the switch statement below should also be updated.

> This code can go away if we don't support rolling upgrade in 0.10.0.

I don't understand, could you please elaborate?